### PR TITLE
Serializer support for maps and sets

### PR DESCRIPTION
**Pull request by hermanventer**
Map, Set, WeakMap and WeakSet can now be serialized.

### CODE REVIEW

**Comment by hermanventer** [comment]
Resolves issues #490 #491 #492

File: src/serializer/serializer.js [hunk 1]
```diff
@@ -750,6 +750,100 @@ export class Serializer {
...
+    let arrayValue = t.arrayExpression(elems);
+    return t.newExpression(
+      this.preludeGenerator.memoizeReference(kind), [arrayValue]);
+  }
+
+  _serializeValueSet(name: string, val: ObjectValue, reasons: Array<string>): BabelNodeExpression {
+    let kind = val.getKind();
+    let elems = [];
+
+    let entries = val.$SetData;
```
Comment (NTillmann) [inline src/serializer/serializer.js]: no need to fetch .$SetData here, you do it again below for "Set".

File: src/serializer/serializer.js [hunk 2]
```diff
@@ -750,6 +750,100 @@ export class Serializer {
...
+      invariant(kind === "WeakSet");
+      entries = val.$WeakSetData;
+    }
+    invariant(entries !== undefined);
+    let len = entries.length;
+
+    for (let i = 0; i < len; i++) {
+      let entry = entries[i];
+      if (entry === undefined) continue;
+      let mightHaveBeenDeleted = entry.mightHaveBeenDeleted();
```
Comment (NTillmann) [inline src/serializer/serializer.js]: I don't quite see how an entry ".mightHaveBeenDeleted()". Is this getting triggered by one of the test cases? If not, and you can't think of one, I suggest to rather have the invariant that !entry.mightHaveBeenDeleted.

File: src/serializer/serializer.js [hunk 3]
```diff
@@ -750,6 +750,100 @@ export class Serializer {
...
+    }
+    invariant(entries !== undefined);
+    let len = entries.length;
+
+    for (let i = 0; i < len; i++) {
+      let entry = entries[i];
+      let key = entry.$Key;
+      let value = entry.$Value;
+      if (key === undefined || value === undefined) continue;
+      let mightHaveBeenDeleted = key.mightHaveBeenDeleted();
```
Comment (NTillmann) [inline src/serializer/serializer.js]: See comment for sets on mightHaveBeenDeleted.

File: src/serializer/serializer.js [hunk 4]
```diff
@@ -750,6 +750,100 @@ export class Serializer {
...
+      entries = val.$WeakMapData;
+    }
+    invariant(entries !== undefined);
+    let len = entries.length;
+
+    for (let i = 0; i < len; i++) {
+      let entry = entries[i];
+      let key = entry.$Key;
+      let value = entry.$Value;
+      if (key === undefined || value === undefined) continue;
```
Comment (NTillmann) [inline src/serializer/serializer.js]: I don't see how this is covered by any test case. Call .clear to trigger it.

**Comment by hermanventer** [comment]
I think there is some more work to be done before we can do a complete test of the deleted cases (see #573). I'd like to get this in, as is, right now since it is blocking us.

**Pull request merged**